torchao/prototype/smoothquant/README.md

```diff
@@ -1,65 +1,61 @@
-python example.py -m MODLE_ID --device=<cuda or cpu> --quant-mode=<dynamic or static> --model-load-path ./quantized_model.pt
+python example.py --repo MODEL_ID --quant smoothquant-0.7 --device cuda
 ```
 
 
 ## Usage of API
 The following APIs are provided:
-- insert_smooth_quant_observer_
 - SmoothQuantConfig
-- save_smooth_quant_recipe (advanced)
-- load_smooth_quant_recipe (advanced)
+- SmoothQuantStep
+- quantize_
 
-`insert_smooth_quant_observer_` inserts observers into the model to be quantized. For example:
+`SmoothQuantConfig` configures applying SmoothQuant to each linear layer of the model. Use it with `torchao.quantization.quantize_`. For example:
 ```python
-insert_smooth_quant_observer_(model, alpha=0.5, quant_mode="dynamic")
-```
-After insertion, run the model for calibration on a certain dataset or (advanced) load a recipe.
+from torchao.prototype.smoothquant import SmoothQuantConfig
+from torchao.prototype.smoothquant.core import SmoothQuantStep
+from torchao.quantization import quantize_
+from torchao.quantization.quant_api import Int8DynamicActivationInt8WeightConfig
 
-`SmoothQuantConfig` configures appliying SmoothQuant to each linear layer of the model. Use it by calling `torchao.quantization.quantize_`. For example:
-```python
-from torchao.prototype.smoothquant import SmoothQuantObservedLinear
-is_observed_linear = lambda m, fqn: isinstance(m, SmoothQuantObservedLinear)
-torchao.quantization.quantize_(model, SmoothQuantConfig(), is_observed_linear)
-```
-`is_observed_linear` is a filter so that we only quantize observed linear layers.
-
-(Advanced) `save_smooth_quant_recipe` and `load_smooth_quant_recipe` saves or loads a recipe for a model.
+# Step 1: Prepare - insert observers
+base_config = Int8DynamicActivationInt8WeightConfig()
+quant_config = SmoothQuantConfig(
+    base_config=base_config,
+    step=SmoothQuantStep.PREPARE,
+    alpha=0.5,
+)
+quantize_(model, quant_config)
 
-A recipe contains smoothing factors and quantization parameters of weights and activation for all linear layers that are to be quantized. For advanced users, these parameters can be saved and modified somehow to produce better accuray, e.g., different alpha for different layers. Users can even leave some linear layers unquantized by deleting these layers in the recipe. Such modifications can be published as a recipe. By loading the recipe, it can be reused and calibration is no longer needed.
-
-To save a recipe, users should insert observers and run calibration first. For example,
-```python
-insert_smooth_quant_observer_(model, alpha=0.5, quant_mode="dynamic")
-for data in dataset_for_calibration:
+# Step 2: Calibration
+for data in calibration_dataset:
     model(data)
-save_smooth_quant_recipe(model, "./smooth_quant_recipe.json")
-```
-To load a recipe, users should insert observers first. For example,
-```python
-insert_smooth_quant_observer_(model)
-load_smooth_quant_recipe(model, "./smooth_quant_recipe.json")
+
+# Step 3: Convert
```

Comment:
I think ideally we can add a tutorial doc for how to save transformer models for vllm/lm-eval as well: https://github.com/pytorch/ao/blob/083361bc3f7addc505a0f994a923f4ae9f54388e/torchao/prototype/awq/example.py#L223-L265

basically: prepare, convert, and then manually set the config step to "prepare_for_loading"
and then upload the model.

after this the model should be able to be used with vllm and lm-eval.

Reply:
sounds good